File: 13. 罗马数字转整数.cpp
#include<iostream>
#include<string>
#include<map>
using namespace std;


class Solution {
public:
	int romanToInt(string s) {
		int length = s.length();
		int num = 0;
		map<char, int> dict;
		dict['I'] = 1;
		dict['V'] = 5;
		dict['X'] = 10;
		dict['L'] = 50;
		dict['C'] = 100;
		dict['D'] = 500;
		dict['M'] = 1000;
		
		for (int i = 0; i < length; i++)
		{
			num += dict[s[i]];
			if (i == 0)
			{
				continue;
			}
			if ((s[i] == 'V' || s[i] == 'X') && s[i-1]=='I') 
			{
				num -= 2;
			}
			else if (((s[i] == 'L' || s[i] == 'C') && s[i - 1] == 'X'))
			{
				num -= 20;
			}
			else if (((s[i] == 'D' || s[i] == 'M') && s[i - 1] == 'C'))
			{
				num -= 200;
			}
		}
		return num;
	}
};

int main()
{
	string temp;
	Solution s = Solution();
	cout << s.romanToInt("IV")<< endl;
	cin >> temp;
	return 0;
}

File: 13.罗马数字转整数.py
class Solution(object):
    def romanToInt(self, str):
        num = 0
        d = {
            "I": 1,
            "V": 5,
            "X": 10,
            "L": 50,
            "C": 100,
            "D": 500,
            "M": 1000
        }
        d2 = {
            "I": ("V", "X"),
            "X": ("L", "C"),
            "C": ("D", "M")
        }
        index = len(str) - 1
        i = 0
        # i 对应 str 中当前字符的下标
        while (i < index):
            # 当前字符在d2中 并且 下一个字符在 对应的集合中, 即属于特殊情况
            if ((str[i] in d2) and (str[i + 1] in d2[str[i]])):
                num += (d[str[i + 1]] - d[str[i]])
                i += 2
            # 一般情况
            else:
                num += d[str[i]]
                i += 1
        # 循环结束可能因为 i+2, i+1 , 第二种情况需要加上字后一个罗马数字对应的数值
        if (i == index):
            num += d[str[-1]]
        return num


if __name__ == "__main__":
    s = Solution()
    print(s.romanToInt("IV"))
